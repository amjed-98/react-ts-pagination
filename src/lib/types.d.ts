import type { CSSProperties } from 'react';

declare global {
  type Color = CSSProperties['color'];
  type BgColor = CSSProperties['backgroundColor'];

  declare type Page = {
    pageNumber: number;
    style: {
      '--color': Color;
      '--background-color': BgColor;
      '--hover-color': Color;
      '--hover-bgColor': BgColor;
    };
    activeClass?: string;
    pageClass?: string;
  };
}
