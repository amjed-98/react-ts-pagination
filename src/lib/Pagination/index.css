#root {
  --color: #a0a7b9;
  --hover-color: white;
  --background-color: transparent;
  --hover-background-color: #8d00d8;
}

.pagination {
  display: flex;
  justify-content: center;
  align-items: baseline;
  padding-block: 4rem;
  margin-top: 1rem;
}

.btn {
  border: none;
  outline: none;
  background-color: transparent;
  cursor: pointer;
  font-size: 1.5rem;
  transition: all 0.1s ease-in-out;
  border-radius: 50%;
  padding: 0.5rem;
  user-select: none;
}

.pages {
  display: flex;
  overflow: hidden;
  max-width: 22rem;
  justify-content: flex-start;
  padding-bottom: 1rem;
  padding-bottom: 1rem;
  gap: 0.8rem;
}

.page {
  font-weight: bold;
  cursor: pointer;
  user-select: none;
  background-color: var(--background-color);
  color: var(--color);
  text-align: center;
  height: 2rem;
  margin: auto 4px;
  display: flex;
  box-sizing: border-box;
  align-items: center;
  justify-content: center;
  letter-spacing: 0.01071em;
  border-radius: 1rem;
  line-height: 1.43;
  font-size: 1rem;
  min-width: 2rem;
}

.active-page {
  background-color: var(--hover-background-color);
  color: var(--hover-color);
}

.page:hover {
  background-color: var(--hover-background-color);
  color: var(--hover-color);
}
