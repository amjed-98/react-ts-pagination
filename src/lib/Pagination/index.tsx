import type { FC, JSXElementConstructor, CSSProperties } from 'react';
import { useCallback, useEffect, useRef } from 'react';
import {
  dispatchEvent,
  getPageRef,
  listenFor,
  scrollToPage,
  handleMouseWheelScroll,
} from '@/lib/utils';
import Button from './components/Button';
import Pages from './components/Pages';
import './index.css';

export type Props = {
  currentPageNumber: number;
  numberOfPages: number;
  onPageChange?: (pageNumber: number, pageRef: HTMLSpanElement | undefined) => void;
  paginationContainerClass?: string;
  pagesContainerClass?: string;
  nextLabel?: string | JSXElementConstructor<Record<string, never>>;
  prevLabel?: string | JSXElementConstructor<Record<string, never>>;
  nextBtnClass?: string;
  prevBtnClass?: string;
  pageStyle?: CSSProperties;
  pageClass?: string;
  activePageStyle?: CSSProperties;
  activePageClass?: string;
};

const Pagination: FC<Props> = (props) => {
  const {
    currentPageNumber,
    numberOfPages,
    pageStyle,
    activePageStyle,
    pageClass = 'page',
    activePageClass = 'active-page',
    paginationContainerClass = 'pagination',
    pagesContainerClass = 'pages',
    nextBtnClass = 'btn',
    prevBtnClass = 'btn',
    onPageChange,
    prevLabel = '❮',
    nextLabel = '❯',
  } = props;

  const pagesRef = useRef<HTMLDivElement>(null);

  const handlePageChange = useCallback((pageNumber: number): void => {
    /*
      ? if pass  onPageChange handler function then will use it
      ? else will use the default handler that is in usePagination hook
    */
    if (onPageChange) {
      const pageRef = getPageRef(pagesRef, pageNumber);

      onPageChange(pageNumber, pageRef);
      pageRef?.scrollIntoView({ behavior: 'smooth' });
    } else dispatchEvent('pageChange', pageNumber);
  }, []);

  useEffect(
    () =>
      listenFor<number>('pageChange', ({ detail: pageNumber }) =>
        scrollToPage(pagesRef, pageNumber),
      ),
    [],
  );

  return (
    <div className={paginationContainerClass} onWheel={handleMouseWheelScroll} role='pagination'>
      <Button
        Label={prevLabel}
        className={prevBtnClass}
        onClick={handlePageChange.bind(null, currentPageNumber - 1)}
      />

      <Pages
        ref={pagesRef}
        currentPageNumber={currentPageNumber}
        numberOfPages={numberOfPages}
        pageStyle={pageStyle}
        activePageStyle={activePageStyle}
        pageClass={pageClass}
        activePageClass={activePageClass}
        handlePageChange={handlePageChange}
        pagesContainerClass={pagesContainerClass}
      />

      <Button
        Label={nextLabel}
        className={nextBtnClass}
        onClick={handlePageChange.bind(null, currentPageNumber + 1)}
      />
    </div>
  );
};

export default Pagination;
