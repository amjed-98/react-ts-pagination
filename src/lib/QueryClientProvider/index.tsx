import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import type { FC, ReactNode } from 'react';

type Props = {
  children: ReactNode | ReactNode[];
};

const queryClient = new QueryClient();

const Provider: FC<Props> = ({ children }) => {
  return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>;
};

export default Provider;
