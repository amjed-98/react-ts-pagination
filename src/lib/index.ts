export { usePagination, useServerPagination } from './hooks';
export { default as Pagination } from './Pagination';
export { default as Provider } from './QueryClientProvider';
