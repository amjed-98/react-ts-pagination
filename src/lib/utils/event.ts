type EventType = keyof DocumentEventMap | 'pageChange' | 'fetch';
type Listener<Payload> = (event: CustomEvent<Payload>) => void;

const listenFor = <Payload>(eventType: EventType, listener: Listener<Payload>): (() => void) => {
  document.addEventListener(eventType, listener as EventListener);

  return () => document.removeEventListener(eventType, listener as EventListener);
};

const dispatchEvent = <Payload>(eventType: EventType, payload: Payload): void => {
  const event = new CustomEvent<Payload>(eventType, { detail: payload });

  document.dispatchEvent(event);
};

const unSubscribe = <Payload>(eventType: EventType, listener: Listener<Payload>): void => {
  document.removeEventListener(eventType, listener as EventListener);
};

export { listenFor, dispatchEvent, unSubscribe };
