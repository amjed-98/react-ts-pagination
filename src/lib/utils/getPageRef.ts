import type { RefObject } from 'react';

type CachedPages = Record<number, HTMLSpanElement | undefined>;

let cachedPages: CachedPages;

const getPageRef = (
  containerRef: RefObject<HTMLDivElement>,
  pageNumber: number,
): HTMLSpanElement | undefined => {
  if (!containerRef.current) return;

  if (!cachedPages) {
    cachedPages = Object.assign({}, containerRef.current.children) as unknown as CachedPages;
  }

  const indexMargin = 1;
  return cachedPages[pageNumber - indexMargin];
};

export default getPageRef;
