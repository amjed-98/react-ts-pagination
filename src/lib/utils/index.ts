export { default as scrollToPage } from './scrollToPage';
export { default as getPageRef } from './getPageRef';
export { dispatchEvent, listenFor, unSubscribe } from './event';
export { default as handleMouseWheelScroll } from './handleMouseWheelScroll';
