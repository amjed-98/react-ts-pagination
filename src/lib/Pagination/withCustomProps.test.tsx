import Pagination, { type Props } from '.';
import { render, renderHook, screen, userEvent } from '@/lib/test_setup';
import { listenFor } from '../utils';
import { useState } from 'react';
import { describe, expect, it, beforeEach } from 'vitest';

const {
  result: {
    current: [currentPageNumber, setCurrentPageNumber],
  },
} = renderHook(() => useState(4));

const props: Props = {
  currentPageNumber,
  numberOfPages: 10,
  nextLabel: 'next',
  prevLabel: 'prev',
  onPageChange: (pageNumber: number) => {
    setCurrentPageNumber(pageNumber);
  },
  nextBtnClass: 'next-btn-class',
  prevBtnClass: 'prev-btn-class',
  pageStyle: { color: 'red' },
  activePageStyle: { color: 'blue' },
  pageClass: 'page-class',
  activePageClass: 'active-page-class',
  pagesContainerClass: 'pages-container-class',
  paginationContainerClass: 'pagination-container-class',
};

const {
  paginationContainerClass,
  nextLabel,
  prevLabel,
  pageClass,
  pageStyle,
  nextBtnClass,
  prevBtnClass,
  activePageClass,
  activePageStyle,
} = props;

const mockListener = vi.fn();
const onPageChangeSpy = vi.spyOn(props, 'onPageChange');

listenFor<number>('pageChange', ({ detail: pageNumber, type }) =>
  mockListener({ pageNumber, type }),
);

describe('Pagination Component  with custom props', () => {
  beforeEach(() => {
    render(<Pagination {...props} />);
  });

  it('should match snapshot', () => {
    const paginationElement = screen.getByRole('pagination');

    expect(paginationElement).toMatchSnapshot();
  });

  it(`should have ${paginationContainerClass} the  class`, () => {
    const paginationElement = screen.getByRole('pagination');

    expect(paginationElement.className).toBe(paginationContainerClass);
  });

  it(`should call onPageChange handler with pageNumber = ${
    currentPageNumber + 1
  } and not dispatch pageChange event when next button clicked`, async () => {
    const nextButton = screen.getByText(nextLabel as string);

    await userEvent.click(nextButton);
    const pageRef = screen.getByText(5);
    expect(onPageChangeSpy).toHaveBeenCalledWith(5, pageRef, props.numberOfPages);

    expect(mockListener).not.toHaveBeenCalled();
  });

  it(`should call onPageChange handler with pageNumber = ${
    currentPageNumber - 1
  } and not dispatch pageChange event when prev button clicked`, async () => {
    const prevButton = screen.getByText(prevLabel as string);

    await userEvent.click(prevButton);
    const pageRef = screen.getByText(3);

    expect(onPageChangeSpy).toHaveBeenCalledWith(3, pageRef, props.numberOfPages);

    expect(mockListener).not.toHaveBeenCalled();
  });

  it(`should give ${nextBtnClass}, ${prevBtnClass} class to buttons`, () => {
    const nextButton = screen.getByText(nextLabel as string);
    const prevButton = screen.getByText(prevLabel as string);

    expect(nextButton.className).toBe(nextBtnClass);
    expect(prevButton.className).toBe(prevBtnClass);
  });

  it(`should give style ${pageStyle?.color}, ${pageClass} class to page element when not active`, () => {
    const pageElement = screen.getByText(1);

    expect(pageElement.style.color).toBe(pageStyle?.color);
    expect(pageElement.className).toBe(pageClass);
  });

  it(`should give style ${activePageStyle?.color}, ${pageClass} ${activePageClass} class to page element when active`, () => {
    const pageElement = screen.getByText(currentPageNumber);

    expect(pageElement.style.color).toBe(activePageStyle?.color);
    expect(pageElement.className).toBe(`${pageClass} ${activePageClass}`);
  });
});
