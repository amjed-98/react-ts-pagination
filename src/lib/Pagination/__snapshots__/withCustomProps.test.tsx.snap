// Vitest Snapshot v1

exports[`Pagination Component  with custom props > should match snapshot 1`] = `
<div
  class="pagination-container-class"
  role="pagination"
>
  <button
    class="prev-btn-class"
  >
    prev
  </button>
  <div
    class="pages-container-class"
  >
    <span
      class="page-class"
      style="color: red;"
    >
      1
    </span>
    <span
      class="page-class"
      style="color: red;"
    >
      2
    </span>
    <span
      class="page-class"
      style="color: red;"
    >
      3
    </span>
    <span
      class="page-class active-page-class"
      style="color: blue;"
    >
      4
    </span>
    <span
      class="page-class"
      style="color: red;"
    >
      5
    </span>
    <span
      class="page-class"
      style="color: red;"
    >
      6
    </span>
    <span
      class="page-class"
      style="color: red;"
    >
      7
    </span>
    <span
      class="page-class"
      style="color: red;"
    >
      8
    </span>
    <span
      class="page-class"
      style="color: red;"
    >
      9
    </span>
    <span
      class="page-class"
      style="color: red;"
    >
      10
    </span>
  </div>
  <button
    class="next-btn-class"
  >
    next
  </button>
</div>
`;
